core/Pos/Util/Util.hs

```diff
@@ -260,6 +261,10 @@ type instance ChannelT (Ether.TaggedTrans tag t m) = ChannelT m
 maybeThrow :: (MonadThrow m, Exception e) => e -> Maybe a -> m a
 maybeThrow e = maybe (throwM e) pure
 
+-- | Fail or return result depending on what is stored in 'Either'.
+eitherToFail :: (MonadFail m, ToString s) => Either s a -> m a
```

Comment:
https://github.com/input-output-hk/cardano-sl/pull/372/files/35aa23e762b8e8ee1a8ef3d7aa69884787d4a5b3#diff-43c2c21b43694f1827d7691b01cdb3bcR78
for example, why did you prefer `MonadError` here?